### PR TITLE
[Doc] Add doc for math module

### PR DESCRIPTION
**Pull request by neozhaoliang**

### CODE REVIEW

File: docs/lang/articles/math/math.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+4. `mat2/mat3/mat4` for 2D/3D/4D floating-point square matrix types.
+
+
+Indeed, these vector/matrix types are created using the two template functions `ti.types.vector()` and `ti.types.matrix()`. For example the `vec2` type is defined in the following way:
+
+```python
+vec2 = ti.types.vector(2, float)
+```
+
+The number of precision bits of these matrix types will be determined by `default_fp` and `default_ip` in your `ti.init()` call. For example, if `ti.init(default_fp=ti.f64)` is called, then the types `vec2/vec3/vec4` and `mat2/mat3/mat4` will all have 64-bit floating-point precision.
```
Comment (Olinaaaloompa) [inline docs/lang/articles/math/math.md:64]:
```suggestion
The number of precision bits of such a type is determined by `default_fp` or `default_ip` in `ti.init()`. For example, if `ti.init(default_fp=ti.f64)` is called, then `vec2/vec3/vec4` and `mat2/mat3/mat4` defined in the Taichi scope all have 64-bit floating-point precision.
```

**Review by Olinaaaloompa — CHANGES_REQUESTED** [review]

File: docs/lang/articles/math/math.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# The math module
+
+
+Taichi implements its own `math` module to support usual mathematical functions plus a few utility functions that are frequently used in computer graphics and simulations. They include:
+
+1. Commonly-used mathematical functions that are analogous to those in Python's built-in `math` module.
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply docs/lang/articles/math/math.md:6]: I'm afraid we should not use "identical" here.

File: docs/lang/articles/math/math.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+2. Small vector and matrix types that are analogous to those in the [OpenGL shading language](https://www.khronos.org/opengl/wiki/OpenGL_Shading_Language)(GLSL).
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply docs/lang/articles/math/math.md:7]: Same. We should use identical here.

File: docs/lang/articles/math/math.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Olinaaaloompa) [reply docs/lang/articles/math/math.md:6]: What about "similar"? No difference between the two modules is mentioned in the following part.

[the remark this answers is not in the document]
Comment (neozhaoliang) [reply docs/lang/articles/math/math.md:6]: The main difference is Taichi's math functions should be called in the Taichi scope, and has different precisions due to the init config and backends.

File: docs/lang/articles/math/math.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+### Relationship between `ti.Vector`, `ti.types.vector` and `ti.math.vec3`
+
++ `ti.Vector` is a function that accepts a 1D array and returns a matrix instance that has only one column. For example, `ti.Vector([1, 2, 3, 4, 5])`.
++ `ti.types.vector` is a function that accepts an integer and a primitive type, and returns a vector type. For example: `vec5f = ti.types.vector(5, float)`. `vec5f` can be used to instantiate 5D vectors or annotate data types in function arguments and struct members:
+    ```python
+    @ti.kernel
+    def test(v: vec5f):
+        print(v.xyz)
+    ```
+    Unlike `ti.Vector`, whose input data must be a 1d array as inputs, vector types created by `ti.types.vector()` have more flexible initialization routines, as we have already explained above.
```
Comment (yuanming-hu) [inline docs/lang/articles/math/math.md:112]:
```suggestion
    Unlike `ti.Vector`, whose input data must be a 1D array, vector types created by `ti.types.vector()` have more flexible ways to initialize, as explained above.
```

File: docs/lang/articles/math/math.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
++ `ti.math.vec3` is created by `vec3 = ti.types.vector(3, float)`.
```
Comment (yuanming-hu) [inline docs/lang/articles/math/math.md:113]:
```suggestion
+ `ti.math.vec3` is simply created by `vec3 = ti.types.vector(3, float)`.
```

File: docs/lang/articles/math/math.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+u = v.xyz  # vec3(1, 2, 3)
+u = v.xxx  # vec3(1, 1, 1)
+u = v.wzyx  # vec4(4, 3, 2, 1)
+u = v.rraa  # vec4(1, 1, 2, 2)
+```
+
+### Relationship between `ti.Vector`, `ti.types.vector` and `ti.math.vec3`
+
++ `ti.Vector` is a function that accepts a 1D array and returns a matrix instance that has only one column. For example, `ti.Vector([1, 2, 3, 4, 5])`.
++ `ti.types.vector` is a function that accepts an integer and a primitive type, and returns a vector type. For example: `vec5f = ti.types.vector(5, float)`. `vec5f` can be used to instantiate 5D vectors or annotate data types in function arguments and struct members:
```
Comment (yuanming-hu) [inline docs/lang/articles/math/math.md:106]:
```suggestion
+ `ti.types.vector` is a function that accepts an integer and a primitive type, and returns a vector type. For example: `vec5f = ti.types.vector(5, float)`. `vec5f` can then be used to instantiate 5D vectors or annotate data types in function arguments and struct members:
```

**Review by yuanming-hu — COMMENTED** [review]
Thanks! Some nits.

File: docs/lang/articles/math/math.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        print(v.xyz)
+    ```
+    Unlike `ti.Vector`, whose input data must be a 1D array, vector types created by `ti.types.vector()` have more flexible ways to initialize, as explained above.
++ `ti.math.vec3` is simply created by `vec3 = ti.types.vector(3, float)`.
+
+
+## GLSL-standard functions
+
+
+Taichi's math module also supports a few [GLSL standard functions](https://registry.khronos.org/OpenGL-Refpages/gl4/index.php), they are implemented in the way that follows the GLSL standard, except that they accept arbitrary vectors and matrices as arguments, and apply to them element-wise. For example:
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/math/math.md:119]:
```suggestion
Taichi's math module also supports a few [GLSL standard functions](https://registry.khronos.org/OpenGL-Refpages/gl4/index.php). These functions follow the GLSL standard, except that they accept arbitrary vectors and matrices as arguments and operate on them element-wise. For example:
```

Please check if I tempered with the intended meaning.

File: docs/lang/articles/math/math.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+Texture support in Taichi is implemented in the `ti.types.texture_types` module.
+
+:::
+
+
+## Complex number arithmetic
+
+Taichi's math module also supports basic complex arithmetic in the form of 2D vectors.
+
+You can use a 2D vector of type `ti.math.vec2` to represent a complex number, the addtion/subtraction of complex numbers are just addtion/subtraction of 2D vectors, and multiplication and division can be performed by calling the two functions `ti.math.cmul` and `ti.math.cdiv`:
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/math/math.md:144]:
```suggestion
You can use a 2D vector of type `ti.math.vec2` to represent a complex number, the addtion/subtraction of complex numbers are just addtion/subtraction of 2D vectors, and multiplication and division can be performed by calling `ti.math.cmul` and `ti.math.cdiv`:
```
Shall we use `ti.math.cmul()` and `ti.math.cdiv()`?

File: docs/lang/articles/math/math.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+:::note
+
+Texture support in Taichi is implemented in the `ti.types.texture_types` module.
+
+:::
+
+
+## Complex number arithmetic
+
+Taichi's math module also supports basic complex arithmetic in the form of 2D vectors.
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/math/math.md:142]: complex arithmetric -> complex operations?

File: docs/lang/articles/math/math.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+:::note
+
+Texture support in Taichi is implemented in the `ti.types.texture_types` module.
+
+:::
+
+
+## Complex number arithmetic
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/math/math.md:140]:
```suggestion
## Complex number operations
```
Correct?

**Review by Vissidarte-Herman — CHANGES_REQUESTED** [review]
LGTM except for some minor editorial requests.

File: docs/lang/articles/math/math.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
+# The math module
```
Comment (Vissidarte-Herman) [inline docs/lang/articles/math/math.md:1]:
```suggestion
# Math Module
```

**Review by Vissidarte-Herman — CHANGES_REQUESTED** [review]

**Review by Vissidarte-Herman — APPROVED** [review]
LGTM

**Review by Olinaaaloompa — APPROVED** [review]

**Pull request merged**